File: switsol/custom_script_project/leave_application.py
from __future__ import unicode_literals
import frappe
from frappe import _
from collections import defaultdict


def on_cancel(self,method=None):
	if self.workflow_state == "Cancelled":
		self.status=self.workflow_state
		frappe.db.set_value("Leave Application", self.name, "status",self.workflow_state)

@frappe.whitelist()
def get_user(doctype, txt, searchfield, start, page_len, filters):
	user = frappe.db.sql("""select approver from `tabEmployee Leave Approver` 
							where user_type = '{0}' 
								and parent = '{1}' 
								and approver like '{txt}'""".format(filters.get('role'),filters.get('employee'),txt= "%%%s%%" % txt),as_list=1)
	return user

@frappe.whitelist()
def get_approver_executor(employee=None):
	if employee:
		employee_doc = frappe.get_doc("Employee",employee)
		approver_counter,executor_counter = 0, 0
		users_dict = defaultdict(list)
		
		for row in employee_doc.leave_approvers:
			if row.user_type == "Approver":
				approver_counter+=1
				if approver_counter == 1:
				 	users_dict[row.user_type].append(row.approver)
				else : users_dict[row.user_type] = ""

			elif row.user_type == "Executor":
				executor_counter+=1
				if executor_counter <= 2:
					users_dict[row.user_type].append(row.approver)
				else : users_dict[row.user_type] = ""
		return users_dict
	


def get_permission_query_conditions(user):
	if not user: user = frappe.session.user
	"""
	get get_permission_query_conditions for approver and leave_executor	
	"""
	if not user == 'Administrator':
		leave_applications = frappe.db.sql("""select name from `tabLeave Application` 
											where approver = '{0}' or leave_executor  = '{0}' or owner = '{0}' """.format(user),as_list=1)
		leave_applications = tuple([application[0].encode('utf-8') for application in leave_applications])
		if leave_applications:
			if len(leave_applications) == 1:
				return "(`tabLeave Application`.name {0} )".format(" = '{0}'".format(leave_applications[0]))
			elif len(leave_applications) > 1:
				return "(`tabLeave Application`.name {0} )".format(" in {0} ".format(leave_applications))
		else:
			return "1=2"

File: switsol/custom_script_project/leave_application.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.ui.form.on("Leave Application", {

	refresh:function(frm){
		set_data(frm)
	},
	approver: function(frm) {
		if(frm.doc.approver){
			frm.set_value("leave_approver_name", frappe.user.full_name(frm.doc.approver));
		}
	},
	validate:function(frm){
		if(user != frm.doc.approver && inList(['Approved by Approver','Rejected by Approver'],frm.doc.workflow_state)) {
			frm.reload_doc();
			frappe.throw(__("Your are not Approver"));
		}
		if((user != frm.doc.leave_executor && user != frm.doc.leave_executor_2) && inList(['Approved','Rejected by Executor','Cancelled'],frm.doc.workflow_state)) {
			frm.reload_doc();
			frappe.throw(__("Your are not Executor"));
		}

		set_data(frm);
	},
	employee:function(frm){
		frappe.call({
			method: "switsol.custom_script_project.leave_application.get_approver_executor",
			args: {
					"employee": frm.doc.employee,
			},
			
			callback: function(r) {
				if(r.message){
					if(r.message['Approver']){
						frm.set_value("approver",r.message['Approver'][0])
						cur_frm.set_df_property("approver", "read_only",1);
					}

					if(r.message['Executor']){
						frm.set_value("leave_executor",r.message['Executor'][0])
						cur_frm.set_df_property("leave_executor", "read_only",1);
					}

					if(r.message['Executor']){
						frm.set_value("leave_executor_2",r.message['Executor'][1])
						cur_frm.set_df_property("leave_executor_2", "read_only",1);
					}
				}
			}
		})
	}
});

cur_frm.fields_dict.approver.get_query = function(doc) {
	return {
		"query": "switsol.custom_script_project.leave_application.get_user",
		"filters": {"role":"Approver","employee":doc.employee}
	}
}

cur_frm.fields_dict.leave_executor.get_query = function(doc) {
	return {
		"query": "switsol.custom_script_project.leave_application.get_user",
		"filters": {"role":"Executor","employee":doc.employee}
	}
}

cur_frm.fields_dict.leave_executor_2.get_query = function(doc) {
	return {
		"query": "switsol.custom_script_project.leave_application.get_user",
		"filters": {"role":"Executor","employee":doc.employee}
	}
}

set_data = function(frm){
	if(frm.doc.approver){
		cur_frm.set_df_property("approver", "read_only",1);
	}
	if(frm.doc.leave_executor){
		cur_frm.set_df_property("leave_executor", "read_only",1);
	}
	if(frm.doc.leave_executor_2){
		cur_frm.set_df_property("leave_executor_2", "read_only",1);
	}
}